finality-aleph/src/sync/handler/mod.rs

```diff
@@ -559,9 +569,175 @@ mod tests {
 
+    fn assert_dangling_branch_required(
+        handler: &Handler<
+            MockBlock,
+            MockPeerId,
+            MockJustification,
+            Backend,
+            Backend,
+            Backend,
+            Backend,
+        >,
+        top: &MockIdentifier,
+        bottom: &MockIdentifier,
+        know_most: HashSet<MockPeerId>,
+    ) {
+        assert!(
+            handler.interest_provider().get(bottom) == Interest::Uninterested,
+            "should not be interested in the bottom"
+        );
+        assert!(
+            handler.interest_provider().get(top)
+                == Interest::Required {
+                    know_most,
+                    branch_knowledge: LowestId(bottom.clone()),
+                },
+            "should require the top"
+        );
+    }
+
+    fn grow_light_branch(
+        handler: &mut Handler<
+            MockBlock,
+            MockPeerId,
+            MockJustification,
+            Backend,
+            Backend,
+            Backend,
+            Backend,
+        >,
+        bottom: &MockIdentifier,
+        length: usize,
+        peer_id: MockPeerId,
+    ) -> Vec<MockHeader> {
+        let branch: Vec<_> = bottom.random_branch().take(length).collect();
+        let top = branch.last().expect("branch should not be empty").id();
+
+        assert!(
+            handler.handle_internal_request(&top).expect("should work"),
+            "should be newly required"
+        );
+        assert!(
+            handler.interest_provider().get(&top)
+                == Interest::Required {
+                    know_most: HashSet::new(),
+                    branch_knowledge: LowestId(top.clone()),
+                },
+            "should be required"
+        );
+
+        let (maybe_id, maybe_error) = handler.handle_request_response(
+            branch
+                .iter()
+                .cloned()
+                .rev()
+                .map(ResponseItem::Header)
+                .collect(),
+            peer_id,
+        );
+
+        assert!(
+            maybe_id.is_none(),
+            "should not create new highest justified"
+        );
+        assert!(maybe_error.is_none(), "should work");
+
+        branch
+    }
+
+    fn create_dangling_branch(
+        handler: &mut Handler<
+            MockBlock,
+            MockPeerId,
+            MockJustification,
+            Backend,
+            Backend,
+            Backend,
+            Backend,
+        >,
+        height: BlockNumber,
+        length: usize,
+        peer_id: MockPeerId,
+    ) -> (MockIdentifier, MockIdentifier) {
+        let bottom = MockIdentifier::new_random(height);
+        let top = grow_light_branch(handler, &bottom, length, peer_id)
+            .last()
+            .expect("branch should not be empty")
+            .id();
+        assert_dangling_branch_required(handler, &top, &bottom, HashSet::from_iter(vec![peer_id]));
+        (bottom, top)
+    }
+
+    fn finalize_light_branch(
```

Comment:
here "light" means "made of `Header`s", I'm not sure if I even want to have such primitive, I hope that this will clarify when I write more complex scenarios